pkg/hubble/relay/pool/manager_test.go

```diff
@@ -0,0 +1,632 @@
+									Address: "192.0.1.1",
+									Type:    peerpb.ChangeNotificationType_PEER_ADDED,
+								}, {
+									Name:    "reachable",
+									Address: "192.0.5.5",
+									Type:    peerpb.ChangeNotificationType_PEER_UPDATED,
+								},
+							}
+							return &testutils.FakePeerNotifyClient{
+								OnRecv: func() (*peerpb.ChangeNotification, error) {
+									i++
+									switch {
+									case i >= len(cns):
+										return nil, nil
+									default:
+										return cns[i], nil
+									}
+								},
+							}, nil
+						},
+						OnClose: func() error {
+							return nil
+						},
+					}, nil
+				},
+			},
+			ccBuilder: FakeClientConnBuilder{
+				OnClientConn: func(target string) (ClientConn, error) {
+					var i int
+					return FakeClientConn{
+						OnGetState: func() connectivity.State {
+							i++
+							if i == 2 {
+								defer func() {
+									close(done)
+								}()
+							}
+							return connectivity.Ready
+						},
+					}, nil
+				},
+			},
+			want: []Peer{
+				{
+					hubblePeer.Peer{
+						Name: "reachable",
+						Address: &net.TCPAddr{
+							IP:   net.ParseIP("192.0.5.5"),
+							Port: defaults.ServerPort,
+						},
+					},
+					FakeClientConn{},
+				},
+			},
+		}, {
+			name: "2 peers added, 1 deleted",
+			pcBuilder: testutils.FakePeerClientBuilder{
+				OnTarget: func() string {
+					return "fake"
+				},
+				OnClient: func() (peerTypes.Client, error) {
+					return &testutils.FakePeerClient{
+						OnNotify: func(_ context.Context, _ *peerpb.NotifyRequest, _ ...grpc.CallOption) (peerpb.Peer_NotifyClient, error) {
+							i := -1
+							cns := []*peerpb.ChangeNotification{
+								{
+									Name:    "one",
+									Address: "192.0.1.1",
+									Type:    peerpb.ChangeNotificationType_PEER_ADDED,
+								}, {
+									Name:    "two",
+									Address: "192.0.1.2",
+									Type:    peerpb.ChangeNotificationType_PEER_ADDED,
+								}, {
+									Name:    "one",
+									Address: "192.0.1.1",
+									Type:    peerpb.ChangeNotificationType_PEER_DELETED,
+								},
+							}
+							return &testutils.FakePeerNotifyClient{
+								OnRecv: func() (*peerpb.ChangeNotification, error) {
+									i++
+									switch {
+									case i >= len(cns):
+										return nil, nil
+									case i == len(cns)-1:
+										defer func() {
+											close(done)
+										}()
+										fallthrough
+									default:
+										return cns[i], nil
+									}
+								},
+							}, nil
+						},
+						OnClose: func() error {
+							return nil
+						},
+					}, nil
+				},
+			},
+			ccBuilder: FakeClientConnBuilder{
+				OnClientConn: func(target string) (ClientConn, error) {
+					return nil, nil
+				},
+			},
+			want: []Peer{
+				{
+					hubblePeer.Peer{
+						Name: "two",
+						Address: &net.TCPAddr{
+							IP:   net.ParseIP("192.0.1.2"),
+							Port: defaults.ServerPort,
+						},
+					},
+					nil,
+				},
+			},
+		}, {
+			name: "PeerClientBuilder errors out",
```

Comment:
Maybe I'm missing something, but it seems we are not checking if these errors are actually thrown somewhere? We just assert that the peer list is empty?

Would it be possible to grep for the error messages in the logger?

Reply:
Indeed. My goal with this test was mostly to make sure this does not lead to a crash. I actually think that it would make sense to allow passing a logger as an option so one could also use a customer logger ([`logrus.FieldLogger`](https://godoc.org/github.com/sirupsen/logrus#FieldLogger) is an interface) for tests that would log to a buffer such that one can assert on the log output.

Reply:
I've updated the tests for test the logger's output.